src/rust/engine/engine_pyo3/src/externs/interface/fs.rs

```diff
@@ -74,6 +86,135 @@ fn match_path_globs(
   })
 }
 
+// -----------------------------------------------------------------------------
+// Digest
+// -----------------------------------------------------------------------------
+
+#[pyclass]
+#[derive(Clone)]
+struct PyDigest(Digest);
+
+#[pymethods]
+impl PyDigest {
+  #[new]
+  fn __new__(fingerprint: &str, serialized_bytes_length: usize) -> PyResult<Self> {
+    let fingerprint = Fingerprint::from_hex_string(fingerprint)
+      .map_err(|e| PyValueError::new_err(format!("Invalid digest hex: {}", e)))?;
+    Ok(Self(Digest::new(fingerprint, serialized_bytes_length)))
+  }
+
+  #[getter]
+  fn fingerprint(&self) -> String {
+    self.0.hash.to_hex()
+  }
+
+  #[getter]
+  fn serialized_bytes_length(&self) -> usize {
+    self.0.size_bytes
+  }
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyDigest {
+  fn __richcmp__(&self, other: PyRef<PyDigest>, op: CompareOp) -> Py<PyAny> {
+    let py = other.py();
+    match op {
+      CompareOp::Eq => (self.0 == other.0).into_py(py),
+      CompareOp::Ne => (self.0 != other.0).into_py(py),
+      _ => py.NotImplemented(),
+    }
+  }
+
+  fn __hash__(&self) -> u64 {
+    self.0.hash.prefix_hash()
+  }
+
+  fn __repr__(&self) -> String {
+    format!("Digest('{}', {})", self.0.hash.to_hex(), self.0.size_bytes)
+  }
+}
+
+// -----------------------------------------------------------------------------
+// Snapshot
+// -----------------------------------------------------------------------------
+
+#[pyclass]
+struct PySnapshot(Snapshot);
+
+#[pymethods]
+impl PySnapshot {
+  #[new]
+  fn __new__() -> Self {
+    Self(Snapshot::empty())
+  }
+
+  #[classmethod]
+  fn _create_for_testing(
+    _cls: &PyType,
+    py_digest: PyDigest,
+    files: Vec<String>,
+    dirs: Vec<String>,
+  ) -> Self {
+    let snapshot = unsafe { Snapshot::create_for_testing_ffi(py_digest.0, files, dirs) };
```

Comment:
Is there a safe alternative to this use of `unsafe`? If not, add a comment marked `SAFETY: ` why this is safe.

Reply:
This function can also be tagged `#[cfg(test)]` so it is only compiled into test binaries.